src/loan.rs

```diff
@@ -60,21 +60,177 @@ static COVENANT_PK: &str = "03b9b6059008e3576aad58e05a3a3e37133b05f68cda8535ec09
 /// `covenant_pk`, which is only used to verify that the transaction
 /// data on the witness stack matches the transaction which triggered
 /// the call.
-fn collateral_descriptor(
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+struct CollateralContract {
+    descriptor: Descriptor<PublicKey>,
     borrower_pk: PublicKey,
     lender_pk: PublicKey,
-    timelock: u64,
-    repayment_output: TxOut,
-) -> Result<Descriptor<PublicKey>> {
-    let repayment_output = serialize(&repayment_output).to_hex();
-    let desc = Descriptor::<elements::bitcoin::PublicKey>::from_str(
-        &(format!(
-            "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
-            COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
-        )),
-    )?;
-
-    Ok(desc)
+    timelock: u32, // TODO: Model timelocks as u32
+}
+
+impl CollateralContract {
+    /// Fill in the collateral contract template with the provided arguments.
+    fn new(
+        borrower_pk: PublicKey,
+        lender_pk: PublicKey,
+        timelock: u64,
+        repayment_output: TxOut,
+    ) -> Result<Self> {
+        let repayment_output = serialize(&repayment_output).to_hex();
+        let descriptor = Descriptor::<elements::bitcoin::PublicKey>::from_str(
+            &(format!(
+                "elcovwsh({},or_i(and_v(v:pk({}),after({})),and_v(v:pk({}),outputs_pref({}))))",
+                COVENANT_PK, lender_pk, timelock, borrower_pk, repayment_output,
+            )),
+        )
+        .context("invalid collateral output descriptor")?;
```

Comment:
Can we ever hit this? May be worth `expect`ing.

Reply:
I thought about it too. I think the only problematic value is the `COVENANT_PK` constant: if it changes and it's no longer a correctly serialized public key then parsing will fail. But we can safely assume that won't happen if we introduce the test that checks that `COVENANT_PK` is the public key corresponding to `COVENANT_SK`, like you suggested [here](https://github.com/comit-network/baru/pull/16#discussion_r672851279) and I forgot to do, woops.

Reply:
Done here: https://github.com/comit-network/baru/pull/25/commits/127b5d825f9fde75ca5d4f34b72aaf46ebcaa20a